src/ol_infrastructure/components/services/vault.py

```diff
@@ -134,3 +137,81 @@ def __init__(
                 default_ttl=db_config.default_ttl
             )
         self.register_outputs({})
+
+class OLVaultAWSSecretsEngineConfig(BaseModel):
+    app_name: Text
+    access_key: Text
+    default_lease_ttl_seconds: int = SIX_MONTHS
+    max_lease_ttl_seconds: int = SIX_MONTHS
+    description: Text
+    secret_key: Text
+    path: Text
+    policy_documents: Dict
+    credential_type: str = 'iam_user'
+
+
+class OLVaultAWSSecretsEngine(ComponentResource):
+
+    def __init__(
+            self,
+            engine_config: OLVaultAWSSecretsEngineConfig,
+            opts: ResourceOptions = None
+    ):
+        super().__init__('ol:services:VaultAWSSecretsEngine', engine_config.app_name, None, opts)
+
+        self.aws_secrets_engine = aws.SecretBackend(
+            # TODO verify app_name exists based on Apps class in ol_types
+            f'aws-{engine_config.app_name}',
+            access_key=engine_config.access_key,
+            secret_key=engine_config.secret_key,
+            path=f'aws-{engine_config.app_name}'
+        )
+
+        for role_name, policy in engine_config.policy_documents.items():
+            aws.SecretBackendRole(
+                role_name,
+                backend=f'aws-{engine_config.app_name}',
+                credential_type='iam_user',
+                name=role_name,
+                policy_document=json.dumps(policy),
+                opts=ResourceOptions(depends_on=[self.aws_secrets_engine])
```

Comment:
This class should have a baseline `opts` object that sets `parent=self` and gets passed to all resources created within the scope of the component resource.

Reply:
Don't think I fully understand your argument

Reply:
So, the pattern I have been using is to create a top level `opts` object at the beginning of the `__init__` method (e.g. [here](https://github.com/mitodl/ol-infrastructure/blob/sar%2Fvault_backend/src/ol_infrastructure/components/aws/database.py#L143))

And then further down for each child resource pass that top level opts object like [here](https://github.com/mitodl/ol-infrastructure/blob/sar%2Fvault_backend/src/ol_infrastructure/components/aws/database.py#L150)

This creates a dependency structure between the component object and the child resources to ensure that everything gets created and cleaned up as a unit.